### PR TITLE
Clean up ALPN when the stream is stopped

### PR DESCRIPTION
**Pull request by raboof**
Refs #2462

Manually tested by running Http2ServerTest, doing a
`curl http://localhost:9001` (speaking HTTP to the HTTPS port)
and inspecting the Jetty ALPN map with a debugger.

### CODE REVIEW

File: akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala [hunk 1]
```diff
@@ -139,16 +139,31 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
...
         engine.setUseClientMode(false)
         Http2AlpnSupport.applySessionParameters(engine, httpsContext.firstSession)
         Http2AlpnSupport.enableForServer(engine, setChosenProtocol)
       }
       val tls = TLS(() ⇒ createEngine, _ ⇒ Success(()), IgnoreComplete)
 
+      val removeEngineOnTerminate: BidiFlow[ByteString, ByteString, ByteString, ByteString, NotUsed] = {
+        implicit val ec = fm.executionContext
+        BidiFlow.fromFlows(
+          Flow.fromFunction(identity),
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala]:
```suggestion
          Flow[ByteString]
```

:) Also below?

File: akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala [hunk 2]
```diff
@@ -139,16 +139,31 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
...
+          Flow
+            .fromFunction[ByteString, ByteString](identity)
+            .watchTermination()((n, fd) ⇒ {
+              fd.onComplete(_ ⇒ eng.foreach(Http2AlpnSupport.cleanupForServer))
+              n
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala]: Is it even needed to return a new mat value? Or maybe `Unit` is just fine.

**Comment by ignasi35** [comment]
If it's the `TLS` stage that's [responsible for creating the `SSLEngine`](https://github.com/akka/akka-http/blob/8f95ab0e4e7c50fdc43833b70b10eec840d98553/akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala#L148) instance, shouldn't it be the one responsible for the cleanup?

This approach works but requires adding a very thin stage that seems a bit out of place.

Adding a `cleanup: () => Try[Unit]` argument in `TLS` and implementing the thin stage there would keep things in place.

**Comment by ignasi35** [comment]
But, yeah, adding an argument to TLS isn't great (new release of `akka/akka`, more API,...).

I'm fine with this solution

**Pull request merged**